lib/created_message_queue_handler.ts

```diff
@@ -100,71 +110,95 @@ export async function handleMessage(
-          : undefined,
-        fiscalCode: retrievedMessage.fiscalCode,
-        id: toNonEmptyString(ulid()).get,
-        kind: "INewNotification",
-        messageId: retrievedMessage.id
-      };
-
-      // save the Notification
-      const result = await notificationModel.create(
-        notification,
-        notification.messageId
-      );
+  if (errorOrMaybeProfile.isLeft) {
+    // query failed
+    return left(ProcessingError.TRANSIENT);
+  }
 
-      if (result.isRight) {
-        // save succeeded, return the saved Notification
-        return right(result.right);
-      } else {
-        // saved failed, fail with a transient error
-        // TODO: we could check the error to see if it's actually transient
-        return left(ProcessingError.TRANSIENT);
+  // query succeeded, we may have a profile
+  const maybeProfile = errorOrMaybeProfile.right;
+
+  // whether the recipient wants us to store the message content
+  const isMessageStorageEnabled =
+    maybeProfile
+      .map(profile => profile.isStorageOfMessageContentEnabled)
+      .getOrElse(false) || false;
+
+  if (isMessageStorageEnabled) {
+    // if the recipient wants to store the messages
+    // we add the content of the message to the message record
+    const errorOrUpdatedMessage = await messageModel.update(
+      retrievedMessage.id,
+      retrievedMessage.fiscalCode,
+      m => {
+        return { ...m, content: messageContent };
       }
-    } else {
-      // query succeeded but no profile was found
-      return left(ProcessingError.NO_ADDRESSES);
+    );
+
+    if (errorOrUpdatedMessage.isLeft) {
+      // we consider errors while updating message as transient
+      return left(ProcessingError.TRANSIENT);
     }
-  } else {
-    // query failed
+  }
+
+  //
+  // attempt to resolve an email notification
+  //
+  const maybeProfileEmail = maybeProfile
+    .flatMap(profile => option(profile.email))
+    .map(email => Tuple2(email, NotificationAddressSource.PROFILE_ADDRESS));
+  const maybeDefaultEmail = defaultAddresses
+    .flatMap(addresses => option(addresses.email))
+    .map(email => Tuple2(email, NotificationAddressSource.DEFAULT_ADDRESS));
+  const maybeEmail = maybeProfileEmail.orElse(() => maybeDefaultEmail);
+  const maybeEmailNotification: Option<
+    INotificationChannelEmail
+  > = maybeEmail.map(({ e1: toAddress, e2: addressSource }) => {
+    return {
+      addressSource,
+      status: NotificationChannelStatus.QUEUED,
+      toAddress
+    };
+  });
+
+  // check whether there's at least a channel we can send the notification to
+  if (maybeEmailNotification.isEmpty) {
+    // no channells to notify the user
+    return left(ProcessingError.NO_ADDRESSES);
+  }
+
+  // create a new Notification object with the configured notification channels
+  const notification: INewNotification = {
+    emailNotification: maybeEmailNotification.isDefined
+      ? maybeEmailNotification.get
+      : undefined,
```

Comment:
what it means 'undefined' here ? (maybe it's worth a comment)

Reply:
= no email notification (there's no email address to send the notification to)

Reply:
yes it looks weird, since there's a check for `maybeEmailNotification.isEmpty` before, this is actually for later, when we have multiple channels, some of them may not be configured (thus the `undefined`)

I'd like to keep it (to avoid the risk of forgetting to add the check later)